pkg/ingester/transfer.go

```diff
@@ -289,3 +477,146 @@ func (i *Ingester) findTargetIngester(ctx context.Context) (*ring.IngesterDesc,
 
 	return &ingesters[0], nil
 }
+
+// unshippedBlocks returns a ulid list of blocks that haven't been shipped
+func unshippedBlocks(dir string) (map[string][]string, error) {
+	userIDs, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	blocks := make(map[string][]string, len(userIDs))
+	for _, user := range userIDs {
+		userID := user.Name()
+		blocks[userID] = []string{} // seed the map with the userID to ensure we xfer the WAL, even if all blocks are shipped
+
+		blockIDs, err := ioutil.ReadDir(filepath.Join(dir, userID))
+		if err != nil {
+			return nil, err
+		}
+
+		m, err := shipper.ReadMetaFile(filepath.Join(dir, userID))
```

Comment:
There may be any race condition reading the meta file while the shipper is shipping a block? Shouldn't we stop the shipper before initiating the transfer?

Reply:
Right, my fault. Thanks for pointing out.